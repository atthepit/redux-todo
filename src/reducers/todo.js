export const ADD_TODO = `ADD_TODO`
export const TOGGLE_TODO = `TOGGLE_TODO`

function todo(state = {}, action) {
  switch (action.type) {
  case ADD_TODO:
    return {
      id: action.id,
      text: action.text,
      complete: false,
    }
  case TOGGLE_TODO:
    if (action.id !== state.id)
      return state

    return {
      ...state,
      complete: !state.complete,
    }
  default:
    return state
  }
}

export default todo
