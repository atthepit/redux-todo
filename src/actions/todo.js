import { ADD_TODO, TOGGLE_TODO } from '../reducers/todo'

export const addTodo = (id, text) => ({
  type: ADD_TODO,
  id,
  text,
})

export const toggleTodo = (id) => ({
  type: TOGGLE_TODO,
  id,
})
