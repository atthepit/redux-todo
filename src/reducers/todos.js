import todo, * as actions from './todo'

function todos(state = [], action) {
  switch (action.type) {
  case actions.ADD_TODO:
    return [...state, todo(undefined, action)]
  case actions.TOGGLE_TODO:
    return state.map(t => todo(t, action))
  default:
    return state
  }
}

export default todos
export const getVisibleTodos = (state, filter) => {
  switch(filter) {
  case `active`:
    return state.filter(t => !t.complete)
  case `completed`:
    return state.filter(t => t.complete)
  default:
    return state
  }
}
