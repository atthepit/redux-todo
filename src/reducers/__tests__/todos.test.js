import todos from '../todos'
import * as actions from '../../actions/todo'

describe('Test todos reducer', () => {
  test('add todo', () => {
    const initialState = undefined
    const action = actions.addTodo(0, 'Hi there!')

    expect(
      todos(initialState, action)
    ).toMatchSnapshot()
  })

  test('add another todo', () => {
    const initialState = [
      { id: 0, text: 'Hello world', complete: false },
      { id: 1, text: 'Hello!', complete: true },
    ]
    const action = actions.addTodo(2, 'Hi there!')

    expect(
      todos(initialState, action)
    ).toMatchSnapshot()
  })

  test('toggle todo', () => {
    const initialState = [
      { id: 0, text: 'Hello world', complete: false },
      { id: 1, text: 'Hello!', complete: true },
    ]
    const action = actions.toggleTodo(1)

    expect(
      // Toggle to false
      todos(initialState, action)
    ).toMatchSnapshot()

    expect(
      // Toggle to true
      todos(initialState, action)
    ).toMatchSnapshot()
  })
})
