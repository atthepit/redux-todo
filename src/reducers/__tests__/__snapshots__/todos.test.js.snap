exports[`Test todos reducer add another todo 1`] = `
Array [
  Object {
    "complete": false,
    "id": 0,
    "text": "Hello world",
  },
  Object {
    "complete": true,
    "id": 1,
    "text": "Hello!",
  },
  Object {
    "complete": false,
    "id": 2,
    "text": "Hi there!",
  },
]
`;

exports[`Test todos reducer add todo 1`] = `
Array [
  Object {
    "complete": false,
    "id": 0,
    "text": "Hi there!",
  },
]
`;

exports[`Test todos reducer toggle todo 1`] = `
Array [
  Object {
    "complete": false,
    "id": 0,
    "text": "Hello world",
  },
  Object {
    "complete": false,
    "id": 1,
    "text": "Hello!",
  },
]
`;

exports[`Test todos reducer toggle todo 2`] = `
Array [
  Object {
    "complete": false,
    "id": 0,
    "text": "Hello world",
  },
  Object {
    "complete": false,
    "id": 1,
    "text": "Hello!",
  },
]
`;
